Add check running service and make extension activation order consistent

lyonsil opened the pull request:
A few notes:
- This is using commands in a different (maybe good, maybe not) way to allow extensions to pass functions to each other. If we want to create a way to do this later we can make the breaking change to do that.
- I am fixing the extension start up order to help resolve some weird start up errors, make activation order consistent, and make the new command to register a check written in TS/JS work consistently. Even if we change from using commands to something else, we probably want to keep extension activation sequential and in order based on dependencies. Added notes to this ticket: https://github.com/paranext/paranext-core/issues/664
- I found while working on this that the VS Code debugger cannot set breakpoints in extension back end code. This needs to be fixed or print-style debugging is the only tool available which is slower and more painful. https://github.com/paranext/paranext-core/issues/989
- I needed to expose the network object service and network object status service on PAPI for the check services to work properly. I don't think this is a problem, but if it is there aren't great alternatives that I know of offhand.
- There is a built-in check that runs automatically if you set DEV_NOISY to true and wait 20 seconds after the test extensions activate. It points out everywhere the word "sheep" is in the book of John.
- When writing the test check, I found that our "surrogate pair friendly" string util functions were running somewhere around 1000x slower than the built-in string functions, so I just used the built-in functions in our check. https://github.com/paranext/paranext-core/issues/992

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/paranext/paranext-core/981)
<!-- Reviewable:end -->

tjcouch-sil reviewed (commented):
Phew! Great stuff :) exciting!

> Reviewed 23 of 23 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/981#-:-O0uYZmp-B_l4AbDB2ED:bt6qyk6)* status: all files reviewed, 8 unresolved discussions (waiting on @lyonsil)

___
*[`.vscode/launch.json` line 45 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0tsYNb2h-jofOB-qbi:-O0tsYNb2h-jofOB-qbj:b-w1snxt) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/.vscode/launch.json#L45)):*
> ```JSON
>         "DEV_NOISY": "false"
>       },
>       "autoAttachChildProcesses": true
> ```

Any luck debugging with this?
___
*[`extensions/src/hello-world/src/main.ts` line 419 at r2](https://reviewable.io/reviews/paranext/paranext-core/981#-O0uXEDM57XwuYNm9q0c:-O0uXEDM57XwuYNm9q0d:bj3qjzu) ([raw file](https://github.com/paranext/paranext-core/blob/9fd00d0ca1aef5b0a2da92f9026fff5e8fc20206/extensions/src/hello-world/src/main.ts#L419)):*
> ```XML
>   papi.webViews.getWebView(htmlWebViewProvider.webViewType, undefined, { existingId: '?' });
>   papi.webViews.getWebView(reactWebViewProvider.webViewType, undefined, { existingId: '?' });
>   papi.webViews.getWebView(reactWebView2Provider.webViewType, undefined, { existingId: '?' });
> ```

Maybe we should make an issue to investigate this problem, figure out why this change fixed the problem, fix the deeper issue if there is one, and possibly get rid of the circular dependency between this and `hello-someone` if you're confident we should stop supporting circular dependencies (at least for now since your new serial activation code doesn't support them)
___
*[`extensions/src/platform-scripture/cspell.json` line 2 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0tuMK7BEKwtPLDLd6r:-O0tuMK7BEKwtPLDLd6s:b-5zqrmb) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/cspell.json#L2)):*
> ```JSON
> {
>   "import": "../../cspell.json"
> ```

This change is pretty similar to deleting `package-lock.json` in the sense that, if we completely nuke this file like this, every time the template changes its `cspell.json` and we update from templates, there will be a merge conflict here. It's an unfortunate situation we're in, but I'd prefer to leave config-related files as close to how they look in the template as possible where appropriate. I'd generally recommend just copying our cspell here over to the template. What do you think? Are there words only relevant to `platform-scripture` that don't seem relevant to the template that you'd like to add here?
___
*[`extensions/src/platform-scripture/src/main.ts` line 66 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0tvlq2BFf5kMoC3p9z:-O0tvlq2BFf5kMoC3pA-:bnizrja) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/src/main.ts#L66)):*
> ```XML
>   );
> 
>   await checkHostingService.initialize();
> ```

I wonder if checks actually belong in their own extension independent of `platform-scripture` since they are kinda a different thing altogether and also since we may be further genericizing them in the future (if that's still the plan) so they work with other kinds of projects than Scripture.
___
*[`extensions/src/platform-scripture/src/checks/check-aggregating.service.ts` line 74 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0u-ywi82N9Sy9uukwe:-O0u-ywi82N9Sy9uukwf:bddifkv) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/src/checks/check-aggregating.service.ts#L74)):*
> ```XML
>   if (!checkRunners) {
>     ranRefresh = true;
>     refreshCheckRunners();
> ```

Do you need to `await` this?
___
*[`extensions/src/platform-scripture/src/checks/check-aggregating.service.ts` line 169 at r2](https://reviewable.io/reviews/paranext/paranext-core/981#-O0uY3hh1297xSE0Oowf:-O0uY3hh1297xSE0Oowg:b-s0ghj6) ([raw file](https://github.com/paranext/paranext-core/blob/9fd00d0ca1aef5b0a2da92f9026fff5e8fc20206/extensions/src/platform-scripture/src/checks/check-aggregating.service.ts#L169)):*
> ```XML
>   }
> 
>   async getCheckResults(): Promise<CheckRunResult[]> {
> ```

It looks like maybe you're intentionally never running a `notifyUpdate` for check results here. Possibly consider not even making this a data provider data type (either add `@papi.dataProviders.decorators.ignore` or whatever it is or change the name so it doesn't have `get` at the start).
___
*[`extensions/src/platform-scripture/src/checks/check-hosting.service.ts` line 180 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0u5JCL4PKhLen_-Wjl:-O0u5JCL4PKhLen_-Wjm:b-spfk1r) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/src/checks/check-hosting.service.ts#L180)):*
> ```XML
>   }
> 
>   async rebuildResults(): Promise<void> {
> ```

Should this run `notifyUpdate` for `CheckResults`? I will admit I skimmed through the comments in your changes to `platform-scripture.d.ts`, so I didn't completely digest every bit of it. Seemed a bit confusing, but maybe it has to do with this not having a `notifyUpdate`.
___
*[`src/extension-host/services/papi-backend.service.ts` line 82 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0uWED87AJb78lUf2n9:-O0uWED87AJb78lUf2nA:b1ze1la) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/src/extension-host/services/papi-backend.service.ts#L82)):*
> ```XML
>   network: papiNetworkService as PapiNetworkService,
>   /** JSDOC DESTINATION networkObjectService */
>   networkObject: networkObjectService as NetworkObjectService,
> ```

We decided we would put an `s` on plural things here a while ago, so I believe this should be `networkObjects` (like `dataProviders`, `webViewProviders`, etc.)
___
*[`src/shared/services/network-object.service.ts` line 541 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0uWajG5brBKcxw8O-S:-O0uWajG5brBKcxw8O-T:b-76g5cz) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/src/shared/services/network-object.service.ts#L541)):*
> ```XML
> 
> // Declare an interface for the object we're exporting so that JSDoc comments propagate
> export interface NetworkObjectService {
> ```

I think we should thin out the things we put on the papi like in the web view service. What do you say we just put `get`, `set`, and `onDidCreateNetworkObject` on the papi? I think `hasKnown` might actually best not be exported at all since it is probably always better to use `get` and I don't think we are using it anywhere.


<!-- Sent from Reviewable.io -->

lyonsil reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/981)* status: all files reviewed, 8 unresolved discussions (waiting on @irahopkinson)

___
*[`.vscode/launch.json` line 45 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0tsYNb2h-jofOB-qbi:-O13BrrVBMDz6U6yHJAR:b-8g47bf) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/.vscode/launch.json#L45)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

Any luck debugging with this?
</blockquote></details>

This might have helped with debugging the extension host itself, so I left it in. It definitely doesn't hurt anything. Documentation on this setting was a little spotty, but it seems like the name is self descriptive, and it is desired for how we debug.

This setting isn't sufficient for (re?)enabling debugging extensions themselves, though, because VSCode still can't resolve breakpoints within extensions even with this on. One issue is that source maps don't seem to be generated for extensions, but even after I generated source maps it didn't help. I left all the source map generation changes out of this PR because:
1) It was kind of confusing whether I ought to make changes to the root dir, the extensions dir, and/or an individual extension's dir to do this "right."
2) Since source map generation itself wasn't enough to fix the problem, I didn't want to start generating useless artifacts that get stored in our builds.
___
*[`extensions/src/platform-scripture/cspell.json` line 2 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0tuMK7BEKwtPLDLd6r:-O13D0TS87DrUaaiRSfT:badtg70) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/cspell.json#L2)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

This change is pretty similar to deleting `package-lock.json` in the sense that, if we completely nuke this file like this, every time the template changes its `cspell.json` and we update from templates, there will be a merge conflict here. It's an unfortunate situation we're in, but I'd prefer to leave config-related files as close to how they look in the template as possible where appropriate. I'd generally recommend just copying our cspell here over to the template. What do you think? Are there words only relevant to `platform-scripture` that don't seem relevant to the template that you'd like to add here?
</blockquote></details>

The `cspell.json` file that was here was a large, incomplete list of words compared to the list in the root directory. I think it's fine for the extension template to have the larger list because extensions generally won't be subdirectories of `paranext-core`'s root directory. Since this extension is special in that way, though, I think it's preferable to just reference the root directory for this bundled extension.

@irahopkinson do you have thoughts about this?


<!-- Sent from Reviewable.io -->

lyonsil reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/981)* status: 19 of 23 files reviewed, 7 unresolved discussions (waiting on @irahopkinson and @tjcouch-sil)

___
*[`extensions/src/platform-scripture/src/main.ts` line 66 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0tvlq2BFf5kMoC3p9z:-O1IKLQ7075iAN-Y2H7R:b-7wmx9f) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/src/main.ts#L66)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

I wonder if checks actually belong in their own extension independent of `platform-scripture` since they are kinda a different thing altogether and also since we may be further genericizing them in the future (if that's still the plan) so they work with other kinds of projects than Scripture.
</blockquote></details>

The "how granular should we be with baked in extensions?" question has no complete answer. I'm leaning toward fewer rather than more extensions because with every extension we add we create another instance of the template with another set of things to merge/watch/build/etc. There are no plans and no stated goal to make the platform work for other things, too, and until there is some inkling of that effort I don't think we should make things harder for ourselves.
___
*[`extensions/src/platform-scripture/src/checks/check-aggregating.service.ts` line 74 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0u-ywi82N9Sy9uukwe:-O1IL6xr7Xdcbp7CQrYp:bmfg1ab) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/src/checks/check-aggregating.service.ts#L74)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

Do you need to `await` this?
</blockquote></details>

Yes, good call.
___
*[`src/extension-host/services/papi-backend.service.ts` line 82 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0uWED87AJb78lUf2n9:-O1ILULy9EQMqktEQNVe:bb6un1j) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/src/extension-host/services/papi-backend.service.ts#L82)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

We decided we would put an `s` on plural things here a while ago, so I believe this should be `networkObjects` (like `dataProviders`, `webViewProviders`, etc.)
</blockquote></details>

Done
___
*[`src/shared/services/network-object.service.ts` line 541 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0uWajG5brBKcxw8O-S:-O1IQJr34-A1QVlMcD-d:bb6un1j) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/src/shared/services/network-object.service.ts#L541)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

I think we should thin out the things we put on the papi like in the web view service. What do you say we just put `get`, `set`, and `onDidCreateNetworkObject` on the papi? I think `hasKnown` might actually best not be exported at all since it is probably always better to use `get` and I don't think we are using it anywhere.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

lyonsil reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/981)* status: 17 of 24 files reviewed, 7 unresolved discussions (waiting on @irahopkinson and @tjcouch-sil)

___
*[`extensions/src/hello-world/src/main.ts` line 419 at r2](https://reviewable.io/reviews/paranext/paranext-core/981#-O0uXEDM57XwuYNm9q0c:-O1OlJmJBPk5N-S4_10T:b-aynhqq) ([raw file](https://github.com/paranext/paranext-core/blob/9fd00d0ca1aef5b0a2da92f9026fff5e8fc20206/extensions/src/hello-world/src/main.ts#L419)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

Maybe we should make an issue to investigate this problem, figure out why this change fixed the problem, fix the deeper issue if there is one, and possibly get rid of the circular dependency between this and `hello-someone` if you're confident we should stop supporting circular dependencies (at least for now since your new serial activation code doesn't support them)
</blockquote></details>

I'm not sure about making an issue for this specific thing, but I think we will want/need some issues for start up concerns. I think Sebastian has already been finding some of those things.
___
*[`extensions/src/platform-scripture/src/checks/check-aggregating.service.ts` line 169 at r2](https://reviewable.io/reviews/paranext/paranext-core/981#-O0uY3hh1297xSE0Oowf:-O1OklGHDcwpXm_2Sywx:b7179q0) ([raw file](https://github.com/paranext/paranext-core/blob/9fd00d0ca1aef5b0a2da92f9026fff5e8fc20206/extensions/src/platform-scripture/src/checks/check-aggregating.service.ts#L169)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

It looks like maybe you're intentionally never running a `notifyUpdate` for check results here. Possibly consider not even making this a data provider data type (either add `@papi.dataProviders.decorators.ignore` or whatever it is or change the name so it doesn't have `get` at the start).
</blockquote></details>

I was missing some `notifyUpdate` calls that should have been there.
___
*[`extensions/src/platform-scripture/src/checks/check-hosting.service.ts` line 180 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0u5JCL4PKhLen_-Wjl:-O1Ol69xAenkkBT6c-2w:b6uoqmh) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/src/checks/check-hosting.service.ts#L180)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

Should this run `notifyUpdate` for `CheckResults`? I will admit I skimmed through the comments in your changes to `platform-scripture.d.ts`, so I didn't completely digest every bit of it. Seemed a bit confusing, but maybe it has to do with this not having a `notifyUpdate`.
</blockquote></details>

Indeed it should have had that. It has been added now.


<!-- Sent from Reviewable.io -->

tjcouch-sil reviewed (approved):
Thanks for the reworks!

> Reviewed 2 of 4 files at r3, 5 of 5 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/981#-:-O1TeJEt1cVssl7_KDGF:b-h0rsfq)* status: all files reviewed, 4 unresolved discussions (waiting on @irahopkinson and @lyonsil)

___
*[`.vscode/launch.json` line 45 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0tsYNb2h-jofOB-qbi:-O1SmjK_1RPEHF6hcQao:b6ljm0i) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/.vscode/launch.json#L45)):*
<details><summary><i>Previously, lyonsil (Matt Lyons) wrote…</i></summary><blockquote>

This might have helped with debugging the extension host itself, so I left it in. It definitely doesn't hurt anything. Documentation on this setting was a little spotty, but it seems like the name is self descriptive, and it is desired for how we debug.

This setting isn't sufficient for (re?)enabling debugging extensions themselves, though, because VSCode still can't resolve breakpoints within extensions even with this on. One issue is that source maps don't seem to be generated for extensions, but even after I generated source maps it didn't help. I left all the source map generation changes out of this PR because:
1) It was kind of confusing whether I ought to make changes to the root dir, the extensions dir, and/or an individual extension's dir to do this "right."
2) Since source map generation itself wasn't enough to fix the problem, I didn't want to start generating useless artifacts that get stored in our builds.
</blockquote></details>

Oh weird, I thought I had set it up so they would get built... I'm almost sure I had this working at some point. Oh well :/ I guess it needs a new issue: #989 dunno what priority, so I just put it at on deck high for now
___
*[`extensions/src/hello-world/src/main.ts` line 419 at r2](https://reviewable.io/reviews/paranext/paranext-core/981#-O0uXEDM57XwuYNm9q0c:-O1So-Iq6m4Zbd68xh5-:b-x2qu92) ([raw file](https://github.com/paranext/paranext-core/blob/9fd00d0ca1aef5b0a2da92f9026fff5e8fc20206/extensions/src/hello-world/src/main.ts#L419)):*
<details><summary><i>Previously, lyonsil (Matt Lyons) wrote…</i></summary><blockquote>

I'm not sure about making an issue for this specific thing, but I think we will want/need some issues for start up concerns. I think Sebastian has already been finding some of those things.
</blockquote></details>

Maybe we can loop this into the discussions about lazy loading and what to do with all that stuff. I don't remember if we quite finished that discussion, so I put an item on our discussion doc.
___
*[`extensions/src/platform-scripture/cspell.json` line 2 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0tuMK7BEKwtPLDLd6r:-O1SoTwe1l1yvy5C2zY1:b-8rstu7) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/cspell.json#L2)):*
<details><summary><i>Previously, lyonsil (Matt Lyons) wrote…</i></summary><blockquote>

The `cspell.json` file that was here was a large, incomplete list of words compared to the list in the root directory. I think it's fine for the extension template to have the larger list because extensions generally won't be subdirectories of `paranext-core`'s root directory. Since this extension is special in that way, though, I think it's preferable to just reference the root directory for this bundled extension.

@irahopkinson do you have thoughts about this?
</blockquote></details>

I imagine we probably need a bigger-picture solution to all this config stuff. But, until we have that solution, I suggest that we treat all these repetitive or unnecessary files the same, and I suggest we continue what have been doing, which is leaving them alone. `package-lock.json`, prettier/linting/etc. configs, etc. are all just copied directly into these extension directories even though they could be inherited from the top-level configs (or, particularly in the case of `package-lock.json`, deleted altogether). Removing the contents of these files (deleting or inheriting from top level) means that, every time we make a change to these files in the template, we will have a merge conflict with every one of these extensions. Leaving them the same as or close to the same as how they are in the templates continues our current solution that minimizes merge conflicts. If we change some config at top level of core, we should also make the same change in the templates, then we update from templates to get those changes reflected in the configs in the extensions. It's annoying, and I confess that I don't always do it over just one little config change at top level. It works in bunches almost as well as it works for every change, so I just get lazy. But I'd like to stick with our current direction until we find another direction.

One partial solution we could choose to at least help somewhat would be to thin out `paranext-extension-template` so it doesn't run on its own. It's *just* for using in the multi-extension template. Then we could have these config files linked up to the higher configs in the template itself. But then we still have the same sticky problem with the multi-extension template and this `paranext-core/extensions` folder, but at least it's only one merge conflict to have to deal with instead of one per bundled extension.

Another partial solution would be to make an npm package or something like erb did that we could share between repos.

I'd like this file to be reverted and continue on with this PR (to continue with the strategy we have been implementing so far) and focus on config stuff another time, but I also left this unblocking because you can just leave it if you want (I think we're both on the same page about focusing on config stuff another time haha).
___
*[`extensions/src/platform-scripture/src/main.ts` line 66 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0tvlq2BFf5kMoC3p9z:-O1TYFOn4cbj-VYFMasj:bwc6ech) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/src/main.ts#L66)):*
<details><summary><i>Previously, lyonsil (Matt Lyons) wrote…</i></summary><blockquote>

The "how granular should we be with baked in extensions?" question has no complete answer. I'm leaning toward fewer rather than more extensions because with every extension we add we create another instance of the template with another set of things to merge/watch/build/etc. There are no plans and no stated goal to make the platform work for other things, too, and until there is some inkling of that effort I don't think we should make things harder for ourselves.
</blockquote></details>

Thinking about this more, I can see a possible use case where someone might not want to be running check-related extensions. Or it's even possible that they should be in a paratext extension repo :thinking: but that feels less necessary. I guess we can leave it here and move it another time if I ever convince you that they should be separate if you're still not convinced.

Hmm now that I think about it, where do the actual checks belong? In the same extension? In a separate extension in paratext repo?
___
*[`extensions/src/platform-scripture/src/checks/check-aggregating.service.ts` line 105 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O1TSnj35ST3FjwODJqY:-O1TSnj35ST3FjwODJqZ:b-188b3b) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/src/checks/check-aggregating.service.ts#L105)):*
> ```XML
>   ): Promise<DataProviderUpdateInstructions<CheckRunnerDataTypes>> {
>     await refreshCheckRunners();
>     // Copy the reference in case it changes underneath us
> ```

I wonder if it's worth putting a warning/notice in the tsdoc for `checkRunners` or a comment where you're using it about this since it's not so obvious anymore. Dunno; probably fine with or without.
___
*[`extensions/src/platform-scripture/src/checks/check-aggregating.service.ts` line 169 at r2](https://reviewable.io/reviews/paranext/paranext-core/981#-O0uY3hh1297xSE0Oowf:-O1TUzoT58RHg1g9dvRj:b-8j1b5x) ([raw file](https://github.com/paranext/paranext-core/blob/9fd00d0ca1aef5b0a2da92f9026fff5e8fc20206/extensions/src/platform-scripture/src/checks/check-aggregating.service.ts#L169)):*
<details><summary><i>Previously, lyonsil (Matt Lyons) wrote…</i></summary><blockquote>

I was missing some `notifyUpdate` calls that should have been there.
</blockquote></details>

Great rework! Love the updates. I think they make everything a lot clearer and easier to read. Thanks!
___
*[`extensions/src/platform-scripture/src/checks/check-hosting.service.ts` line 0 at r4](https://reviewable.io/reviews/paranext/paranext-core/981#-O1TVS0bB9-5ABoMtO9t:-O1TVS0bB9-5ABoMtO9u:bccxotl) ([raw file](https://github.com/paranext/paranext-core/blob/9ee756f1f98a2485c3b56dca67fe8df925ab80c4/extensions/src/platform-scripture/src/checks/check-hosting.service.ts#L0)):*
The name of this file and the aggregating file (and the class names to some extent) confuse me. They seem a bit incongruous with what is in the files. This hosting name especially seems confusing (I can't tell the difference between aggregating and hosting without looking in the files even though I'm comfortable with the concepts in this PR). Maybe consider renaming to something more self-explanatory like:

`check-aggregating.service.ts` -> `check-data-provider-engine.model.ts` (or maybe rename both the name and the class itself to be something a bit more self-explanatory like `CheckAggregatorDataProviderEngine` or something

`check-hosting.service.ts` -> `check-runner-engine.model.ts` (maybe even `extension-host-check-runner-engine.model.ts` to be more explicit about what it is doing? I dunno; not so convinced about that)

I'm sorry I didn't mention this before... I think I planned to and then forgot or maybe got lazy :grimacing:
___
*[`extensions/src/platform-scripture/src/checks/check-hosting.service.ts` line 133 at r4](https://reviewable.io/reviews/paranext/paranext-core/981#-O1TdTdb79zndgRtPgdy:-O1TdTdb79zndgRtPgdz:b76ge1t) ([raw file](https://github.com/paranext/paranext-core/blob/9ee756f1f98a2485c3b56dca67fe8df925ab80c4/extensions/src/platform-scripture/src/checks/check-hosting.service.ts#L133)):*
> ```XML
>         if (!this.subscribedProjects.has(projectId)) {
>           // We just need something to tell us when project data changes
>           await pdp.subscribeBookUSFM(
> ```

I think this is returning an unsubscriber that isn't getting run on disposal. Dunno; maybe that's a larger issue that we should try to fix some other way...? But I thought you did this in the aggregator. Not sure.
___
*[`extensions/src/platform-scripture/src/checks/check-hosting.service.ts` line 138 at r4](https://reviewable.io/reviews/paranext/paranext-core/981#-O1TatKz1WEmLAjtGsVA:-O1TatKz1WEmLAjtGsVB:b-64cm3w) ([raw file](https://github.com/paranext/paranext-core/blob/9ee756f1f98a2485c3b56dca67fe8df925ab80c4/extensions/src/platform-scripture/src/checks/check-hosting.service.ts#L138)):*
> ```XML
>               // Ideally we'd want to check if the text that changed is inside of an active range
>               // This isn't available to subscriptions right now, so just rebuild everything
>               this.rebuildResults(projectId);
> ```

Looks like this is an unawaited promise - what do you think of making this function `async` and putting a `try/catch` around this with a nice error message?


<!-- Sent from Reviewable.io -->

lyonsil reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/981)* status: all files reviewed, 4 unresolved discussions (waiting on @irahopkinson and @tjcouch-sil)

___
*[`extensions/src/platform-scripture/src/main.ts` line 66 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0tvlq2BFf5kMoC3p9z:-O1Th_oU91gE0qKY0GxD:b-x6sg31) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/src/main.ts#L66)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

Thinking about this more, I can see a possible use case where someone might not want to be running check-related extensions. Or it's even possible that they should be in a paratext extension repo :thinking: but that feels less necessary. I guess we can leave it here and move it another time if I ever convince you that they should be separate if you're still not convinced.

Hmm now that I think about it, where do the actual checks belong? In the same extension? In a separate extension in paratext repo?
</blockquote></details>

I'm assuming the PT9 "look-a-like" checks written in TS will either be added to this same extension or put into an extension that holds them all. I haven't put a lot of thought into that, either.

Where I am with this higher level point is that we don't have a philosophy that clearly lays out which extensions should hold which things. In the absence of a clear approach, I'd prefer to do the thing that requires less maintenance than the thing that requires more. If we come up with a clear philosophy that works well, then we can reevaluate this and other bits of logic.

In other situations I've seen the "fine grained approach" to packaging code and dependencies play out in practice. Taking that to an extreme is mightily painful and often gets reversed. I'm not suggesting we go completely in the other direction, either, and have one big, fat extension with everything for PT10 in it. I just lean toward "chunky" extensions vs. "tiny" extensions.
___
*[`extensions/src/platform-scripture/src/checks/check-aggregating.service.ts` line 105 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O1TSnj35ST3FjwODJqY:-O1TjTEsDtSxySxLSLVU:bwdw5hn) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/src/checks/check-aggregating.service.ts#L105)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

I wonder if it's worth putting a warning/notice in the tsdoc for `checkRunners` or a comment where you're using it about this since it's not so obvious anymore. Dunno; probably fine with or without.
</blockquote></details>

I don't really follow your concern here. What would the warning/notice say? What should check authors be concerned about?
___
*[`extensions/src/platform-scripture/src/checks/check-hosting.service.ts` line 133 at r4](https://reviewable.io/reviews/paranext/paranext-core/981#-O1TdTdb79zndgRtPgdy:-O1Tmh2LERNrgPX8pvkY:bseka24) ([raw file](https://github.com/paranext/paranext-core/blob/9ee756f1f98a2485c3b56dca67fe8df925ab80c4/extensions/src/platform-scripture/src/checks/check-hosting.service.ts#L133)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

I think this is returning an unsubscriber that isn't getting run on disposal. Dunno; maybe that's a larger issue that we should try to fix some other way...? But I thought you did this in the aggregator. Not sure.
</blockquote></details>

I did think about that, but I didn't think it mattered because:
1) The check hosting service itself will never be deactivated except when the process goes away, I think.
2) If the PDP goes away this subscription doesn't matter, and other things will be cleaned up a few lines later under `pdp.onDidDispose`.

Does this assuage your concern?
___
*[`extensions/src/platform-scripture/src/checks/check-hosting.service.ts` line  at r4](https://reviewable.io/reviews/paranext/paranext-core/981#-O1TVS0bB9-5ABoMtO9t:-O1Tjqe-4peg9dIfmPF2:b-da95td) ([raw file](https://github.com/paranext/paranext-core/blob/9ee756f1f98a2485c3b56dca67fe8df925ab80c4/extensions/src/platform-scripture/src/checks/check-hosting.service.ts#L)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

The name of this file and the aggregating file (and the class names to some extent) confuse me. They seem a bit incongruous with what is in the files. This hosting name especially seems confusing (I can't tell the difference between aggregating and hosting without looking in the files even though I'm comfortable with the concepts in this PR). Maybe consider renaming to something more self-explanatory like:

`check-aggregating.service.ts` -> `check-data-provider-engine.model.ts` (or maybe rename both the name and the class itself to be something a bit more self-explanatory like `CheckAggregatorDataProviderEngine` or something

`check-hosting.service.ts` -> `check-runner-engine.model.ts` (maybe even `extension-host-check-runner-engine.model.ts` to be more explicit about what it is doing? I dunno; not so convinced about that)

I'm sorry I didn't mention this before... I think I planned to and then forgot or maybe got lazy :grimacing:
</blockquote></details>

Sorry about that. You have no idea how many times I renamed things trying to make it cleaner, and I guess it's still not quite where it needs to be.

The "hosting service" as currently named is the service that owns all checks written in TS/JS and are registered using `platformScripture.registerCheck`. It eliminates the need for writing an entire `ICheckRunner` data provider for a check written in TS/JS, instead giving devs the ability to create a an object of type `Check` which is almost entirely just code that implements the logic of taking project data and finding the intended problems. The "hosting service" is itself an `ICheckRunner` data provider.

The "aggregating service" muxes calls across all `ICheckRunner` data providers so things like our UI only have to talk to a single service to communicate with all `ICheckRunner`s. Note that checks written in any other language cannot use the "hosting service" and instead must implement `ICheckRunner`s to do what they need.

Do you have better naming ideas for these? I can add more details to JSDOCs, too, if it helps.

I tried to avoid using "engine" or "data provider" in the names of either because:
1) They are both data providers themselves, and
2) Neither is an "engine" in the way we've used the term for data providers in general.

In general I think all the `model` files we have in other places only have exported types and maybe a few small (e.g., string, number) exported `const`s. These two files have objects that are registered on the network and require initialization, so I don't think either is really a `model` per se.


<!-- Sent from Reviewable.io -->

tjcouch-sil reviewed (approved):
Three optional things to consider, but <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> other than that! :) thanks!

> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/981#-:-O1XKbkA58tzwav8wsvn:bbu70yf)* status: all files reviewed, 3 unresolved discussions (waiting on @irahopkinson and @lyonsil)

___
*[`extensions/src/platform-scripture/src/main.ts` line 66 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0tvlq2BFf5kMoC3p9z:-O1WtljQ61vvxNpmb4p7:b-dr1uq5) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/src/main.ts#L66)):*
<details><summary><i>Previously, lyonsil (Matt Lyons) wrote…</i></summary><blockquote>

I'm assuming the PT9 "look-a-like" checks written in TS will either be added to this same extension or put into an extension that holds them all. I haven't put a lot of thought into that, either.

Where I am with this higher level point is that we don't have a philosophy that clearly lays out which extensions should hold which things. In the absence of a clear approach, I'd prefer to do the thing that requires less maintenance than the thing that requires more. If we come up with a clear philosophy that works well, then we can reevaluate this and other bits of logic.

In other situations I've seen the "fine grained approach" to packaging code and dependencies play out in practice. Taking that to an extreme is mightily painful and often gets reversed. I'm not suggesting we go completely in the other direction, either, and have one big, fat extension with everything for PT10 in it. I just lean toward "chunky" extensions vs. "tiny" extensions.
</blockquote></details>

With the shift away from lazy loading extensions and toward internal lazy loading, I definitely understand a desire to get chunkier. However, I do think it is worth separating distinct features. Maybe this is something we should discuss at more length. I'll add it to our conversation list.
___
*[`extensions/src/platform-scripture/src/checks/check-aggregating.service.ts` line 105 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O1TSnj35ST3FjwODJqY:-O1WuIrq19ychWI1IHBn:byxtjd) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/src/checks/check-aggregating.service.ts#L105)):*
<details><summary><i>Previously, lyonsil (Matt Lyons) wrote…</i></summary><blockquote>

I don't really follow your concern here. What would the warning/notice say? What should check authors be concerned about?
</blockquote></details>

Oh, sorry, I see it doesn't make you see the file at the same revisions as I was on when I made the comment. Weird. I had in view the previous to the current revision, so I put a comment on the line that says the following:

```typescript
// Copy the reference in case it changes underneath us
```

Don't feel like you need to do anything with this unless you feel it is particularly compelling.
___
*[`extensions/src/platform-scripture/src/checks/check-hosting.service.ts` line 133 at r4](https://reviewable.io/reviews/paranext/paranext-core/981#-O1TdTdb79zndgRtPgdy:-O1XHOWG06x9kxr_nTGU:b-3blag) ([raw file](https://github.com/paranext/paranext-core/blob/9ee756f1f98a2485c3b56dca67fe8df925ab80c4/extensions/src/platform-scripture/src/checks/check-hosting.service.ts#L133)):*
<details><summary><i>Previously, lyonsil (Matt Lyons) wrote…</i></summary><blockquote>

I did think about that, but I didn't think it mattered because:
1) The check hosting service itself will never be deactivated except when the process goes away, I think.
2) If the PDP goes away this subscription doesn't matter, and other things will be cleaned up a few lines later under `pdp.onDidDispose`.

Does this assuage your concern?
</blockquote></details>

Hmm I guess the update event's subscriptions do get cleared out when the pdp is disposed... I think that does mean this isn't a memory leak :) good thinking! Hopefully this case is already covered!
___
*[`extensions/src/platform-scripture/src/checks/check-hosting.service.ts` line  at r4](https://reviewable.io/reviews/paranext/paranext-core/981#-O1TVS0bB9-5ABoMtO9t:-O1Wv5F83nF1NBK_Bod4:b1fvm74) ([raw file](https://github.com/paranext/paranext-core/blob/9ee756f1f98a2485c3b56dca67fe8df925ab80c4/extensions/src/platform-scripture/src/checks/check-hosting.service.ts#L)):*
<details><summary><i>Previously, lyonsil (Matt Lyons) wrote…</i></summary><blockquote>

Sorry about that. You have no idea how many times I renamed things trying to make it cleaner, and I guess it's still not quite where it needs to be.

The "hosting service" as currently named is the service that owns all checks written in TS/JS and are registered using `platformScripture.registerCheck`. It eliminates the need for writing an entire `ICheckRunner` data provider for a check written in TS/JS, instead giving devs the ability to create a an object of type `Check` which is almost entirely just code that implements the logic of taking project data and finding the intended problems. The "hosting service" is itself an `ICheckRunner` data provider.

The "aggregating service" muxes calls across all `ICheckRunner` data providers so things like our UI only have to talk to a single service to communicate with all `ICheckRunner`s. Note that checks written in any other language cannot use the "hosting service" and instead must implement `ICheckRunner`s to do what they need.

Do you have better naming ideas for these? I can add more details to JSDOCs, too, if it helps.

I tried to avoid using "engine" or "data provider" in the names of either because:
1) They are both data providers themselves, and
2) Neither is an "engine" in the way we've used the term for data providers in general.

In general I think all the `model` files we have in other places only have exported types and maybe a few small (e.g., string, number) exported `const`s. These two files have objects that are registered on the network and require initialization, so I don't think either is really a `model` per se.
</blockquote></details>

(I think what you wrote here about the services and their big-picture purpose is great. I would recommend putting a revised version of that at the top of each file or something along those lines and probably also in TSDocs in `platform-scripture.d.ts`. I think it's probably hard for someone to understand the big picture right now without having the context of our lengthy discussions together)

I see what you mean about neither being an engine but more resembling a service; neither of the classes extending `DataProviderEngine` are what is finally exported... :thinking: I guess in some sense it feels more natural to me to think about the two files as really being about the data providers, though. I dunno; probably not that big a deal. I think the weirdness I'm experiencing is considering how these data providers are conceptualized inside and outside of this extension. Inside the extension, they're thought about more as part of a service. But outside, they're thought about as just normal data providers that you get by an id. My brain thinks it's easier to think the same way internally as we would think externally to the extension since that's the api we're providing.

After some thought, I suggest that the data provider names be the same as the file name/service but without the word "service" (and without adding "data provider"). Feels helpful in associating them at a glance while still appropriately representing that they are slightly different things internally and externally. My new suggestion is as follows:

- file name and exported object `checkAggregatingService` -> `checkAggregatorService` and data provider name `platformScripture.checkAggregatingDataProvider` -> `platformScripture.checkAggregator`
  - remove the words "Data Provider" since it's already inherently a data provider and change the last word to a noun. It parallels `platformScripture.extensionHostCheckRunner` this way and is more concise.

- file name and exported object `checkHostingService` -> `extensionHostCheckRunnerService`. Data provider name stays the same `platformScripture.extensionHostCheckRunner`
  - Feels nice to use the same term "check runner" for this file and for what it represents and is consistent with the data provider name, too

I think these changes together help with understanding the difference between the two files and the big picture purpose of the files at a glance.


<!-- Sent from Reviewable.io -->

lyonsil commented:
*[`extensions/src/platform-scripture/cspell.json` line 2 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0tuMK7BEKwtPLDLd6r:-O1YjaMz5W9krJ94nwku:b5h0cyr) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/cspell.json#L2)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

I imagine we probably need a bigger-picture solution to all this config stuff. But, until we have that solution, I suggest that we treat all these repetitive or unnecessary files the same, and I suggest we continue what have been doing, which is leaving them alone. `package-lock.json`, prettier/linting/etc. configs, etc. are all just copied directly into these extension directories even though they could be inherited from the top-level configs (or, particularly in the case of `package-lock.json`, deleted altogether). Removing the contents of these files (deleting or inheriting from top level) means that, every time we make a change to these files in the template, we will have a merge conflict with every one of these extensions. Leaving them the same as or close to the same as how they are in the templates continues our current solution that minimizes merge conflicts. If we change some config at top level of core, we should also make the same change in the templates, then we update from templates to get those changes reflected in the configs in the extensions. It's annoying, and I confess that I don't always do it over just one little config change at top level. It works in bunches almost as well as it works for every change, so I just get lazy. But I'd like to stick with our current direction until we find another direction.

One partial solution we could choose to at least help somewhat would be to thin out `paranext-extension-template` so it doesn't run on its own. It's *just* for using in the multi-extension template. Then we could have these config files linked up to the higher configs in the template itself. But then we still have the same sticky problem with the multi-extension template and this `paranext-core/extensions` folder, but at least it's only one merge conflict to have to deal with instead of one per bundled extension.

Another partial solution would be to make an npm package or something like erb did that we could share between repos.

I'd like this file to be reverted and continue on with this PR (to continue with the strategy we have been implementing so far) and focus on config stuff another time, but I also left this unblocking because you can just leave it if you want (I think we're both on the same page about focusing on config stuff another time haha).
</blockquote></details>

I made this particular change because I added a bunch of words to the `cspell` list to get rid of warnings in the extension, and it seemed a little odd to only make the change to the extension given the terms I was adding.

It feels cleaner to just reference the parent cspell config, but I can revert this and just copy the new root-level cspell config here.


<!-- Sent from Reviewable.io -->

lyonsil reviewed (commented):
Thanks for all the thoughtful feedback!

> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/981#-:-O1Yof3R01lFEsEuZR5n:bwdgghj)* status: 19 of 24 files reviewed, all discussions resolved (waiting on @tjcouch-sil)

___
*[`extensions/src/platform-scripture/src/main.ts` line 66 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0tvlq2BFf5kMoC3p9z:-O1YjJAG8d5iSIUZ-HNb:bnmf6rf) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/src/main.ts#L66)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

With the shift away from lazy loading extensions and toward internal lazy loading, I definitely understand a desire to get chunkier. However, I do think it is worth separating distinct features. Maybe this is something we should discuss at more length. I'll add it to our conversation list.
</blockquote></details>

Sounds good - we can always do some refactoring
___
*[`extensions/src/platform-scripture/src/checks/check-aggregating.service.ts` line 105 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O1TSnj35ST3FjwODJqY:-O1XuD-5-Iqs3EkNMqNR:b-927veh) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/src/checks/check-aggregating.service.ts#L105)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

Oh, sorry, I see it doesn't make you see the file at the same revisions as I was on when I made the comment. Weird. I had in view the previous to the current revision, so I put a comment on the line that says the following:

```typescript
// Copy the reference in case it changes underneath us
```

Don't feel like you need to do anything with this unless you feel it is particularly compelling.
</blockquote></details>

Ah, got it. Yeah that comment probably doesn't apply anymore with the refactor, so I'll probably leave it as-is.
___
*[`extensions/src/platform-scripture/src/checks/check-hosting.service.ts` line 138 at r4](https://reviewable.io/reviews/paranext/paranext-core/981#-O1TatKz1WEmLAjtGsVA:-O1YiFxvDvb0tblGfO6o:bq2x5pc) ([raw file](https://github.com/paranext/paranext-core/blob/9ee756f1f98a2485c3b56dca67fe8df925ab80c4/extensions/src/platform-scripture/src/checks/check-hosting.service.ts#L138)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

Looks like this is an unawaited promise - what do you think of making this function `async` and putting a `try/catch` around this with a nice error message?
</blockquote></details>

Good idea - done
___
*[`extensions/src/platform-scripture/src/checks/check-hosting.service.ts` line  at r4](https://reviewable.io/reviews/paranext/paranext-core/981#-O1TVS0bB9-5ABoMtO9t:-O1YbfYf2PvRdrMmt_ES:b7foyd6) ([raw file](https://github.com/paranext/paranext-core/blob/9ee756f1f98a2485c3b56dca67fe8df925ab80c4/extensions/src/platform-scripture/src/checks/check-hosting.service.ts#L)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

(I think what you wrote here about the services and their big-picture purpose is great. I would recommend putting a revised version of that at the top of each file or something along those lines and probably also in TSDocs in `platform-scripture.d.ts`. I think it's probably hard for someone to understand the big picture right now without having the context of our lengthy discussions together)

I see what you mean about neither being an engine but more resembling a service; neither of the classes extending `DataProviderEngine` are what is finally exported... :thinking: I guess in some sense it feels more natural to me to think about the two files as really being about the data providers, though. I dunno; probably not that big a deal. I think the weirdness I'm experiencing is considering how these data providers are conceptualized inside and outside of this extension. Inside the extension, they're thought about more as part of a service. But outside, they're thought about as just normal data providers that you get by an id. My brain thinks it's easier to think the same way internally as we would think externally to the extension since that's the api we're providing.

After some thought, I suggest that the data provider names be the same as the file name/service but without the word "service" (and without adding "data provider"). Feels helpful in associating them at a glance while still appropriately representing that they are slightly different things internally and externally. My new suggestion is as follows:

- file name and exported object `checkAggregatingService` -> `checkAggregatorService` and data provider name `platformScripture.checkAggregatingDataProvider` -> `platformScripture.checkAggregator`
  - remove the words "Data Provider" since it's already inherently a data provider and change the last word to a noun. It parallels `platformScripture.extensionHostCheckRunner` this way and is more concise.

- file name and exported object `checkHostingService` -> `extensionHostCheckRunnerService`. Data provider name stays the same `platformScripture.extensionHostCheckRunner`
  - Feels nice to use the same term "check runner" for this file and for what it represents and is consistent with the data provider name, too

I think these changes together help with understanding the difference between the two files and the big picture purpose of the files at a glance.
</blockquote></details>

Thanks for the thoughtful suggestions! I renamed things as you suggest and tried to improve the JSDOC comments. I agree these are clearer than what I had.


<!-- Sent from Reviewable.io -->

tjcouch-sil reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :tada: :tada: :tada:

> Reviewed 5 of 5 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/981#-:-O1bL1EuDOEsrJADzM6s:bwupo5h)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved

___
*[`extensions/src/platform-scripture/cspell.json` line 2 at r1](https://reviewable.io/reviews/paranext/paranext-core/981#-O0tuMK7BEKwtPLDLd6r:-O1bEXfq2uOHXvqXlQCu:b5x9xgs) ([raw file](https://github.com/paranext/paranext-core/blob/ba27f2dc603ac583c44846f3f4e58f0162bd2427/extensions/src/platform-scripture/cspell.json#L2)):*
<details><summary><i>Previously, lyonsil (Matt Lyons) wrote…</i></summary><blockquote>

I made this particular change because I added a bunch of words to the `cspell` list to get rid of warnings in the extension, and it seemed a little odd to only make the change to the extension given the terms I was adding.

It feels cleaner to just reference the parent cspell config, but I can revert this and just copy the new root-level cspell config here.
</blockquote></details>

I agree that it does feel cleaner. Let's hope we find a good solution soon :grimacing: thanks for the adjustment.
___
*[`extensions/src/platform-scripture/src/checks/check-hosting.service.ts` line  at r4](https://reviewable.io/reviews/paranext/paranext-core/981#-O1TVS0bB9-5ABoMtO9t:-O1bIT8e55luFovhGazA:b-ucfknh) ([raw file](https://github.com/paranext/paranext-core/blob/9ee756f1f98a2485c3b56dca67fe8df925ab80c4/extensions/src/platform-scripture/src/checks/check-hosting.service.ts#L)):*
<details><summary><i>Previously, lyonsil (Matt Lyons) wrote…</i></summary><blockquote>

Thanks for the thoughtful suggestions! I renamed things as you suggest and tried to improve the JSDOC comments. I agree these are clearer than what I had.
</blockquote></details>

Thanks!


<!-- Sent from Reviewable.io -->
